### ISSUE TITLE
[docs] Typed client registration issue

### ISSUE THREAD

**Issue by denisenko93** [dotnet/runtime#96308]
### Description

When registering several implementations of typed httpclient on a common interface, the httpclient that was registered last will be resolved as a dependency

### Reproduction Steps

```CSharp
var hostBuilder = Host.CreateApplicationBuilder(args);

hostBuilder.Services.AddHttpClient<ITypedClient,TypedClient1>(x => x.BaseAddress = new Uri("http://example.com"));
hostBuilder.Services.AddHttpClient<ITypedClient,TypedClient2>(x => x.BaseAddress = new Uri("https://github.com"));

var host = hostBuilder.Build();

await host.StartAsync();

var clients = host.Services.GetServices<ITypedClient>();

foreach (ITypedClient typedClient in clients)
{
    Console.WriteLine($"Client: {typedClient.GetType()}. Base url: {typedClient.Client.BaseAddress}");
}

await host.StopAsync();

interface ITypedClient
{
    HttpClient Client { get; }
}

record TypedClient1(HttpClient Client): ITypedClient;
record TypedClient2(HttpClient Client): ITypedClient;
```

### Expected behavior

In console:
```
Client: TypedClient1. Base url: https://github.com/
Client: TypedClient2. Base url: https://github.com/
```

### Actual behavior

In console:
```
Client: TypedClient1. Base url: http://example.com/
Client: TypedClient2. Base url: https://github.com/
```

### Configuration

.NET SDK:
 Version:           8.0.100
 Commit:            57efcf1350
 Workload version:  8.0.100-manifests.6c33ef20

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100\

### Other information

Registering httpclient without an interface and additionally registering an interface for each implementation fixes this problem
```CSharp
hostBuilder.Services.AddHttpClient<TypedClient1>(x => x.BaseAddress = new Uri("http://example.com"));
hostBuilder.Services.AddHttpClient<TypedClient2>(x => x.BaseAddress = new Uri("https://github.com"));

hostBuilder.Services.AddTransient<ITypedClient>(x => x.GetService<TypedClient1>());
hostBuilder.Services.AddTransient<ITypedClient>(x => x.GetService<TypedClient2>());
```

**Comment by MMaximus111** [comment]
Yes, I faced the same problem, not logical behavior...

**Comment by CarnaViire** [comment]
TL;DR: This is, unfortunately, expected behavior. Typed clients use Named clients "under the hood", and in this case they both would reuse the same Named client with a name "ITypedClient". See the docs for [`AddHttpClient<TClient,TImplementation>(IServiceCollection, Action<HttpClient>)`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-8.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-action((system-net-http-httpclient)))). The workaround is to specify different names explicitly e.g. by using the overload [`AddHttpClient<TClient,TImplementation>(IServiceCollection, String, Action<HttpClient>)`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-8.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-string-system-action((system-net-http-httpclient)))).

-----

As mentioned in the [docs](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-8.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-action((system-net-http-httpclient)))), a Typed client implicitly registers and links to a Named client. The client name, unless explicitly provided, will be set to the type name of `TClient`. Note that this would be the first one from the pair in case `AddHttpClient<TClient,TImplementation>` overloads are used.

Therefore, registering a Typed client technically does two separate things:
1) registers a Named client (in a simple default case, the name is `typeof(TClient).Name`), and
2) registers a Transient service using the `TClient` or `TClient,TImplementation` provided.

The following statement
```c#
services.AddHttpClient<ITypedClient, TypedClient1>(c => c.BaseAddress = new Uri("http://example.com"));
```

would be technically the same as
```c#
services.AddHttpClient("ITypedClient", c => c.BaseAddress = new Uri("http://example.com")) // register Named client
    .AddTypedClient<ITypedClient, TypedClient1>(); // link the Named client to a Typed client
```

and even -- in this simple case -- same as
```c#
services.AddHttpClient("ITypedClient", c => c.BaseAddress = new Uri("http://example.com")); // register Named client

// register plain Transient and link it to the Named client
Services.AddTransient<ITypedClient, TypedClient1>(s =>
    new TypedClient1(
        s.GetRequiredService<IHttpClientFactory>().CreateClient("ITypedClient")));
```

The non-obvious part of HttpClientFactory is that, by design, registering/configuring a Named client with the same name several times would just _append_ the configuration actions to the list of existing ones. This is the approach used by configuration APIs like `ConfigureNamedOptions`.

This is mostly useful for handler configurations, e.g. adding some custom handler to a client defined externally, substituting something for tests etc, but it works for `HttpClient` actions as well. For example, this
```c#
services.AddHttpClient("ITypedClient", c =>
    {
        c.BaseAddress = new Uri("http://example.com");
        c.DefaultRequestHeaders.UserAgent.ParseAdd("HttpClient/8.0");
    });
```

and this
```c#
services.AddHttpClient("ITypedClient", c => c.BaseAddress = new Uri("http://example.com"));
services.AddHttpClient("ITypedClient")
    .ConfigureHttpClient(c => c.DefaultRequestHeaders.UserAgent.ParseAdd("HttpClient/8.0"));
```

will result in a client configured in the same way (both BaseAddress and DefaultRequestHeaders are set).

This enables linking a Typed client to an already defined Named client, and/or, linking several Typed clients to a single Named client. It is more obvious when overloads with a name parameter are used:
```c#
services.AddHttpClient("LogClient", c => c.BaseAddress = new Uri(LogServerAddress));

services.AddHttpClient<FooLogger>("LogClient");
services.AddHttpClient<BarLogger>("LogClient");
```

In your case, where you _don't_ need to reuse the Named client, you need to explicitly specify different names for them:
```c#
hostBuilder.Services.AddHttpClient<ITypedClient,TypedClient1>("example", x => x.BaseAddress = new Uri("http://example.com"));
hostBuilder.Services.AddHttpClient<ITypedClient,TypedClient2>("github", x => x.BaseAddress = new Uri("https://github.com"));
```

The bottom line is, this behavior, though it might seem confusing, was designed to be this way. It was designed quite some time ago, and now we might agree or not agree with some decisions made by people back then. But changing things like that would be too big of a breaking change. I'm not sure it would be worth it.

----

We should add explicit description of this behavior to the conceptual docs. I believe they don't mention it now, only API docs do, and it's not extensive. I think we might just reuse the explanations above.

**Comment by denisenko93** [comment]
I suggest using a pair of interface, implementation as a name for the named client. This will allow registration of several implementations with a common interface.
Then such an implementation will be analogous to:
```CSharp
services.AddHttpClient("ITypedClient,TypedClient1", c => c.BaseAddress = new Uri("http://example.com")) // register Named client
    .AddTypedClient<ITypedClient, TypedClient1>();
```

**Comment by CarnaViire** [comment]
> I suggest using a pair of interface, implementation as a name for the named client.

As I mentioned before, this would be a breaking change. If there are people who rely on the current behavior (that the named client is reused), their code will _stop working_ (depending on the actual config, either all or most of typed clients which relied on that would unexpectedly start using an unconfigured HttpClient instead of a configured one). And this will be _extremely hard_ to catch and to root cause, because there will be no compile time errors and no startup errors.

In general, instead of breaking the old behavior, it is preferable to add some _opt-in mechanism_, so that the people who _want_ to remove the old behavior can _explicitly_ set that in the code. In this particular case, this mechanism already exists in the form of an overload with the name.

And the workaround is even smaller than the one you've used above, because, as I mentioned, you can use [AddHttpClient<TClient,TImplementation>(IServiceCollection, String, Action<HttpClient>)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-8.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-string-system-action((system-net-http-httpclient)))). So it is almost the same one-liner as the original registration, plus opt-in to a different name.

```c#
services.AddHttpClient<ITypedClient, TypedClient1>("ITypedClient,TypedClient1", c => c.BaseAddress = new Uri("http://example.com"));
```

**Comment by denisenko93** [comment]
I believe that this breaking change is only for users who register httpclient in a similar way
```CSharp
services.AddHttpClient<ITypedClient, TypedClient1>( c => c.BaseAddress = new Uri("http://example.com"));
services.AddHttpClient("ITypedClient", c => c.DefaultRequestHeaders.UserAgent.ParseAdd("HttpClient/8.0"));
```
But such an approach in registration indicates that the user is based on the internal implementation of AddHttpClient, which is not described anywhere. That is, on undocumented behavior. Therefore, I believe that this behavior can be changed

**Comment by CarnaViire** [comment]
> which is not described anywhere. That is, on undocumented behavior.

This is _not_ an undocumented behavior. It is documented here: [AddHttpClient<TClient,TImplementation>(IServiceCollection, Action<IServiceProvider,HttpClient>)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.httpclientfactoryservicecollectionextensions.addhttpclient?view=dotnet-plat-ext-8.0#microsoft-extensions-dependencyinjection-httpclientfactoryservicecollectionextensions-addhttpclient-2(microsoft-extensions-dependencyinjection-iservicecollection-system-action((system-iserviceprovider-system-net-http-httpclient)))). Below is an extract from this doc:

### AddHttpClient<**_TClient_**,TImplementation>(IServiceCollection, Action<IServiceProvider,HttpClient>)

Adds the IHttpClientFactory and related services to the IServiceCollection and configures a binding between the TClient type and **_a named HttpClient_**. The client name will be set to **_the type name of TClient._**

**Comment by CarnaViire** [comment]
Covered in https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory-troubleshooting#typed-client-has-the-wrong-httpclient-injected

(And more specifically, https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory-troubleshooting#different-typed-clients-are-registered-on-a-common-interface)
